compiler/src/parsing/parser.mly

```diff
@@ -0,0 +1,706 @@
+  | ENUM TYPEID id_vec? data_constructors { Dat.variant ~loc:(to_loc $loc) (mkstr $loc($2) $2) (Option.value ~default:[] $3) $4 }
+  | RECORD TYPEID id_vec? data_labels { Dat.record ~loc:(to_loc $loc) (mkstr $loc($2) $2) (Option.value ~default:[] $3) $4 }
+
+prim1_expr:
+  | NOT non_assign_expr { Exp.apply ~loc:(to_loc $loc) (mkid_expr $loc($1) ["!"]) [$2] }
+
+paren_expr:
+  | lparen expr rparen { $2 }
+
+app_expr:
+  | left_accessor_expr lparen lseparated_list(comma, expr) comma? rparen { Exp.apply ~loc:(to_loc $loc) $1 $3 }
+
+// These are all inlined to carry over their precedence.
+%inline plus_op:
+  | PLUS { "+" }
+%inline plusplus_op:
+  | PLUSPLUS { "++" }
+%inline dash_op:
+  | DASH { "-" }
+%inline star_op:
+  | STAR { "*" }
+%inline slash_op:
+  | SLASH { "/" }
+%inline percent_op:
+  | PERCENT { "%" }
+%inline is_op:
+  | IS { "is" }
+%inline isnt_op:
+  | ISNT { "isnt" }
+%inline eqeq_op:
+  | EQEQ { "==" }
+%inline noteq_op:
+  | NOTEQ { "!=" }
+%inline caret_op:
+  | CARET { "^" }
+%inline lcaret_op:
+  | LCARET { "<" }
+%inline llcaret_op:
+  | LCARETLCARET { "<<" }
+%inline rcaret_op:
+  | RCARET { ">" }
+%inline rrcaret_op:
+  | RCARET RCARET { ">>" }
+%inline rrrcaret_op:
+  | RCARET RCARET RCARET { ">>>" }
+%inline lesseq_op:
+  | LESSEQ { "<=" }
+%inline greatereq_op:
+  | GREATEREQ { ">=" }
+%inline amp_op:
+  | AMP { "&" }
+%inline ampamp_op:
+  | AMPAMP { "&&" }
+%inline pipe_op:
+  | PIPE { "|" }
+%inline pipepipe_op:
+  | PIPEPIPE { "||" }
+%inline pluseq_op:
+  | PLUSEQ { "+" }
+%inline dasheq_op:
+  | DASHEQ { "-" }
+%inline stareq_op:
+  | STAREQ { "*" }
+%inline slasheq_op:
+  | SLASHEQ { "/" }
+%inline percenteq_op:
+  | PERCENTEQ { "%" }
+
+%inline infix_op:
+  | plus_op
+  | dash_op
+  | star_op
+  | slash_op
+  | percent_op
+  | is_op
+  | isnt_op
+  | eqeq_op
+  | plusplus_op
+  | noteq_op
+  | caret_op
+  | lcaret_op
+  | llcaret_op
+  | rcaret_op
+  | rrcaret_op
+  | rrrcaret_op
+  | lesseq_op
+  | greatereq_op
+  | amp_op
+  | ampamp_op
+  | pipe_op
+  | pipepipe_op {$1}
+
+prefix_op:
+  | NOT { "!" }
+
+primitive_:
+  | ASSERT { "assert" }
+  | THROW { "throw" }
+  | FAIL { "fail" }
+
+special_op:
+  | infix_op | prefix_op {$1}
+
+%inline special_id:
+  | lparen special_op rparen { $2 }
+
+%inline modid:
+  | lseparated_nonempty_list(dot, TYPEID) { $1 }
+
+non_modid:
+  | ID
+  | special_id { [$1] }
+
+id:
+  | modid dot non_modid { mkid (List.append $1 $3) (to_loc $loc) }
+  | modid %prec DOT { (mkid $1) (to_loc $loc) }
+  | non_modid { (mkid $1) (to_loc $loc) }
+
+simple_id:
+  | ID { (mkid [$1]) (to_loc $loc) }
+
+type_id:
+  | lseparated_nonempty_list(dot, TYPEID) { (mkid $1) (to_loc $loc) }
+
+id_expr:
+  // Force any following colon to cause a shift
+  | id %prec COLON { Exp.ident ~loc:(to_loc $loc) $1 }
+
+simple_expr:
+  | const { let (exp, loc) = $1 in Exp.constant ~loc:(to_loc loc) exp }
```

Comment:
Another one of these strange looking ones. I understood it more this 2nd time I encountered it.

Reply:
Let's do the same thing as above where we don't introduce a let-binding

Reply:
```suggestion
  | const { Exp.constant ~loc:(to_loc (snd $1)) (fst $1) }
```